src/Microsoft.AspNetCore.Mvc.TagHelpers/InputTagHelper.cs

```diff
@@ -254,15 +254,44 @@ protected string GetInputType(ModelExplorer modelExplorer, out string inputTypeH
 
         private void GenerateCheckBox(ModelExplorer modelExplorer, TagHelperOutput output)
         {
-            if (typeof(bool) != modelExplorer.ModelType)
+            var isValidBool = false;
+            if (modelExplorer.ModelType == typeof(bool))
             {
-                throw new InvalidOperationException(Resources.FormatInputTagHelper_InvalidExpressionResult(
-                    "<input>",
-                    ForAttributeName,
-                    modelExplorer.ModelType.FullName,
-                    typeof(bool).FullName,
-                    "type",
-                    "checkbox"));
+                isValidBool = true;
+            }
+            else if (modelExplorer.ModelType == typeof(string))
+            {
+                bool potentialBool;
+                isValidBool = bool.TryParse(modelExplorer.Model?.ToString() ?? string.Empty, out potentialBool);
+            }
+
+            if (!isValidBool)
+            {
+                if (modelExplorer.ModelType == typeof(string))
+                {
+                    if (modelExplorer.Model != null)
+                    {
+                        throw new InvalidOperationException(Resources.FormatInputTagHelper_InvalidStringResult(
+                            "<input>",
+                            ForAttributeName,
+                            modelExplorer.ModelType.FullName,
+                            typeof(string).FullName,
+                            typeof(bool).FullName,
+                            "checkbox"));
+                    }
+                }
+                else if (modelExplorer.ModelType != typeof(bool))
+                {
+                    throw new InvalidOperationException(Resources.FormatInputTagHelper_InvalidExpressionResult(
+                       "<input>",
+                       ForAttributeName,
+                       modelExplorer.ModelType.FullName,
+                       typeof(bool).FullName,
+                       typeof(string).FullName,
+                       "type",
+                       "checkbox"));
+                }
+                
```

Comment:
I wasn't clear enough in my earlier comments. This block and the one above can be combined. No need for the `isValidBool` variable. Then you're only checking the types once.